File: LeetCode Problems/1001 to 1100/1008. Construct Binary Search Tree from Preorder Traversal/1008. Construct Binary Search Tree from Preorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private TreeNode buildBST(TreeNode root, int data) {
        if (root == null)
            root = new TreeNode(data);
        else if (root.val > data)
            root.left = buildBST(root.left, data);
        else
            root.right = buildBST(root.right, data);
        return root;
    }
    public TreeNode bstFromPreorder(int[] preorder) {
        TreeNode root = null;
        for (int i = 0; i < preorder.length; i++) root = buildBST(root, preorder[i]);
        return root;
    }
}

File: LeetCode Problems/1001 to 1100/1008. Construct Binary Search Tree from Preorder Traversal/1008. Construct Binary Search Tree from Preorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode *buildBST(TreeNode *root, int data) {
        if (root == nullptr)
            root = new TreeNode(data);
        else if (root->val > data)
            root->left = buildBST(root->left, data);
        else
            root->right = buildBST(root->right, data);
        return root;
    }
    TreeNode *bstFromPreorder(vector<int> &preorder) {
        TreeNode *root = nullptr;
        for (int i = 0; i < preorder.size(); i++)
            root = buildBST(root, preorder[i]);
        return root;
    }
};